OpenRA.Mods.Common/Traits/World/Pathfinder/Graph.cs

```diff
@@ -0,0 +1,187 @@
+		// is the three cells ahead, plus the two cells to the side, which we cannot exclude without knowing if
+		// the cell directly between them and our parent is passable.
+		static readonly CVec[][] DirectedNeighbors = {
+			new[] { new CVec(-1, -1), new CVec(0, -1), new CVec(1, -1), new CVec(-1, 0), new CVec(-1, 1) },
+			new[] { new CVec(-1, -1), new CVec(0, -1), new CVec(1, -1) },
+			new[] { new CVec(-1, -1), new CVec(0, -1), new CVec(1, -1), new CVec(1, 0), new CVec(1, 1) },
+			new[] { new CVec(-1, -1), new CVec(-1, 0), new CVec(-1, 1) },
+			CVec.Directions,
+			new[] { new CVec(1, -1), new CVec(1, 0), new CVec(1, 1) },
+			new[] { new CVec(-1, -1), new CVec(-1, 0), new CVec(-1, 1), new CVec(0, 1), new CVec(1, 1) },
+			new[] { new CVec(-1, 1), new CVec(0, 1), new CVec(1, 1) },
+			new[] { new CVec(1, -1), new CVec(1, 0), new CVec(-1, 1), new CVec(0, 1), new CVec(1, 1) },
+		};
+
+		public ICollection<GraphConnection> GetConnections(CPos position)
+		{
+			var previousPos = this.cellInfo[position].PreviousPos;
+
+			var dx = position.X - previousPos.X;
+			var dy = position.Y - previousPos.Y;
+			var index = dy * 3 + dx + 4;
+
+			var validNeighbors = new LinkedList<GraphConnection>();
+			var directions = DirectedNeighbors[index];
+			for (int i = 0; i < directions.Length; i++)
+			{
+				var neighbor = position + directions[i];
+				var movementCost = GetCostToNode(neighbor, directions[i]);
+				if (movementCost != InvalidNode)
+					validNeighbors.AddLast(new GraphConnection(position, neighbor, movementCost));
+			}
+
+			return validNeighbors;
+		}
+
+		int GetCostToNode(CPos destNode, CVec direction)
+		{
+			int movementCost;
+			if (this.mobileInfo.CanEnterCell(
+				this.World as World,
+				this.Actor as Actor,
+				destNode,
+				out movementCost,
+				this.IgnoredActor as Actor,
+				checkConditions) && !(this.CustomBlock != null && this.CustomBlock(destNode)))
+			{
+				return CalculateCellCost(destNode, direction, movementCost);
+			}
+
+			return InvalidNode;
+		}
+
+		int CalculateCellCost(CPos neighborCPos, CVec direction, int movementCost)
+		{
+			var cellCost = movementCost;
+
+			if (direction.X * direction.Y != 0)
+				cellCost = (int)(cellCost * Constants.Sqrt2);
+
+			if (CustomCost != null)
+				cellCost += CustomCost(neighborCPos);
+
+			// directional bonuses for smoother flow!
+			if (LaneBias != 0)
+			{
+				var ux = neighborCPos.X + (InReverse ? 1 : 0) & 1;
+				var uy = neighborCPos.Y + (InReverse ? 1 : 0) & 1;
+
+				if ((ux == 0 && direction.Y < 0) || (ux == 1 && direction.Y > 0))
+					cellCost += LaneBias;
+
+				if ((uy == 0 && direction.X < 0) || (uy == 1 && direction.X > 0))
+					cellCost += LaneBias;
+			}
+
+			return cellCost;
+		}
+
+		bool disposed;
+		public void Dispose()
+		{
+			if (disposed)
+				return;
+
+			disposed = true;
+
+			CellInfoLayerManager.Instance.PutBackIntoPool(this.cellInfo);
+			this.cellInfo = null;
+
+			GC.SuppressFinalize(this);
+		}
+
+		~Graph() { Dispose(); }
```

Comment:
Not really, it's a special kind of Dispose to reuse CellLayers. It was already implemented on bleed, I just moved the declaration.